### ISSUE TITLE
cog-execute does not execute PutLink inside TrueLink in some situation.

### ISSUE THREAD

**Issue by astroseger** [opencog/atomspace#2312]
Let's consider the following example (simplified version of [examples/atomspace/random-choice.scm](https://github.com/opencog/atomspace/blob/master/examples/atomspace/random-choice.scm))

```lisp
(use-modules (opencog) (opencog exec))

; Initialize count to zero
(State (Anchor "sum") (Number 0))

;  define increment "function"
(Define (DefinedPredicate "inc")
                (True (Put (State (Anchor "sum") (Variable "$x")) 
                 (Plus (Number 1) (Get (State (Anchor "sum") (Variable "$y")))))))

; In the following, cog-evaluate! will execute PutLink and increment sum
(cog-evaluate! (DefinedPredicate "inc"))
(cog-execute! (Get (State (Anchor "sum") (Variable "$x"))))

; In the following, cog-execute! will do nothing
(cog-execute! (DefinedPredicate "inc"))
(cog-execute! (Get (State (Anchor "sum") (Variable "$x"))))


; But cog-execute! directly with TrueLink (not via DefinedPredicate) will work.
(cog-execute!  (True (Put (State (Anchor "sum") (Variable "$x")) 
                       (Plus (Number 1) (Get (State (Anchor "sum") (Variable "$y")))))))
(cog-execute! (Get (State (Anchor "sum") (Variable "$x"))))

```

So cog-execute! will not execute PutLink inside TrueLink if it is hidden inside DefinedPredicate.

**Comment by linas** [comment]
Yes, seems to be a real bug. I can look at it maybe in a few days or so. This might be easy to fix, maybe. Or maybe not. Care to try fixing it?

The reason it might not be easy to fix is that we are in the middle of a (year-long) transition in the way `cog-execute!` where it is taking on more and more roles responsibilities. Thus, changes like this sometimes require larger cleanups of old code.  For a newcomer, it might be hard to tell apart ugly old code that should be removed from the "new style" that should be created.  So you can try to fix it, but I might say, "oh no that's the wrong way..."

**Comment by astroseger** [comment]
@linas 
Which is exact semantic for cog-execute! cog-evaluate! ? I mean  the target semantic which will be after implementation of ongoing changes. Actually it is very confusing that in ```recursive-loop.scm``` and ```random-choice.scm``` examples ( from [examples/atomspace](https://github.com/opencog/atomspace/tree/master/examples/atomspace)) cog-evaluate is used for modifying atomspace (increase counter for example).

**Comment by linas** [comment]
Some history then:  Both functions can run code that "changes the atomspace". The (original) difference between them was that `cog-evaluate!` only returned truth values (TV's) and `cog-execute!` only returned Atoms.  This  was back in the day when the TV and the Atom hierarchies were completely distinct (and very different).  Since then, both were modified to derive from a single base type: `Value`. Later on, `cog-execute!` was modified to return Value instead of Atom, and it now does this, and mostly it seems to work quite well... usually.  It's still a work-in-progress, mostly done but still incomplete.

Meanwhile, `cog-evaluate!` continues to return only TV's. The way that it invokes the code to run is quite very different than how `cog-execute!` works. In the long run, it is probably wise to make `cog-evaluate!` be a special case of `cog-execute!`. Probably, maybe, not sure, not entirely clear; needs thinking, needs work. It probably should be me looking and thinking about this; but I'm very busy with other things.

So how can one move these two functions closer together? Well, one way is brute-force, stupidly, by mandate, without any thinking. That would be wrong, I don't want that. But if you can think of incremental steps moving in that direction, that I would like.

-----------
The reason that `cog-execute!` on `TrueLink` doesn't work is because there is no C++ class called `TrueLink` and there are no methods `TrueLink::execute()` and `TrueLink::is_exeutable()`. Perhaps, maybe, such a class should exist?  Better yet, maybe there should just be a base class, `EvaluatableLink`, which has `EvaluatableLink::execute()` and `EvaluateLink::is_exeutable()`, and that is enough to make everything work?  If I was doing it, that is what I might try: take a small step in that direction, see if it feels good, or see if there are other, better ways ... if it feels good, if it passes unit tests, then merge...

**Comment by linas** [comment]
For example (and I'm not quite sure this is a good idea, but ...) it could look  like this:
```
class EvaluatableLink {
    bool is_executable() { return true; }
    ValuePtr execute() { return EvaluationLink::do_evaluate(this->atomspace, this->getHandle()); }
};
```
and maybe something like that is enough to fix it!?

**Comment by linas** [comment]
Closing, fixed by #2327
